### DISCUSSION TITLE
No response received, expected at least 8 bytes (0 received)

### DISCUSSION THREAD

**Discussion by JimmyTang999** [pymodbus-dev/pymodbus#2413] [General]
This is very similar to a [post](https://stackoverflow.com/questions/75682169/modbus-error-input-output-modbus-error-invalid-message-no-response-receive) from stack overflow that was never answered like 2 years ago

its working with radzio
![ModbusError](https://github.com/user-attachments/assets/daf6daa3-e276-48e1-9076-8248696cc23a)

just not with python- pymodbus==3.7.4
```

from pymodbus.client import ModbusTcpClient
client = ModbusTcpClient('192.168.1.195', port=26)  # Replace with your server's IP and port
client.connect()
result = client.read_holding_registers(address=4000, count=10, slave=1) 
if result.isError():
    print("Error reading registers:", result)
else:
    print("Register values:", result.registers)
client.close()


```

**Comment by janiversen** [comment]
Please add a debug log, as the issue template showed you how to do, so we can see what happens.

We do not follow stack overflow, and apart from that the library was very different 2 years ago.

The typical problem in this case, is that your device do not respond. A modbus socket packet is at least 8 bytes, but we received 0 within the timeout period. Are you sure slave=1 is correct for your device ?

**Reply by JimmyTang999** [reply]
The debug log

```
DEBUG:pymodbus.logging:Connection to Modbus server established. Socket ('192.168.1.225', 64136)
DEBUG:root:Connected to the Modbus server.
DEBUG:pymodbus.logging:Current transaction state - IDLE
DEBUG:pymodbus.logging:Running transaction 1
DEBUG:pymodbus.logging:SEND: 0x0 0x1 0x0 0x0 0x0 0x6 0x1 0x3 0xf 0xa0 0x0 0xa
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Transaction failed. (Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)) 
DEBUG:pymodbus.logging:Processing: 
DEBUG:pymodbus.logging:Getting transaction 1
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:root:Read result: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)
DEBUG:pymodbus.logging:Connection to Modbus server established. Socket ('192.168.1.225', 64137)
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 2
DEBUG:pymodbus.logging:SEND: 0x0 0x2 0x0 0x0 0x0 0x6 0x1 0x3 0xf 0xa0 0x0 0xa
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:No response received to unbounded read !!!!
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 
DEBUG:pymodbus.logging:Processing: 
DEBUG:pymodbus.logging:Getting transaction 2
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
```
salve is indeed 1 this is what I used in Radzio for Device ID when it worked.

**Reply by janiversen** [reply]
But what about the difference that the other app uses rtu over tcp/ip and you program uses socket ?

As the log clearly shows there are no response, so the above difference is likely the cause. Please refer to the documentation how to use rtu over tcp/ip.

**Reply by JimmyTang999** [reply]
Probably you are correct but i'm not sure where to look in the docs

1) looking here https://pymodbus.readthedocs.io/en/dev/source/library/pymodbus.html#pymodbus.FramerType.RTU
-i see where RTU_OVER_TCP is supported
-i noticed the FramerType class
tried something like this
from pymodbus.framer import ModbusRtuFramer
client = ModbusTcpClient('192.168.1.195', port=26,type=RTU)

after looking at __init__.py within framer folder, i'm guessing i need this
from pymodbus.framer import FramerRTU

still not really sure how to use it though

**Reply by janiversen** [reply]
Why not look at the examples we provide included in the documentation, that would be very simple, instead of looking at init etc...but of course that is your choice.Start by reading the README would give you all the pointers you need.

I really cannot help you more, I told you what the problem most likely is, where to look for explanations and examples.

**Reply by JimmyTang999** [reply]
heres the WireShark dump when i run from radzio
```
0020   01 c3 dd 0a 00 1a 73 5d af f3 00 03 6b 2e 50 18   ......s]....k.P.
0030   fa 72 85 17 00 00 01 03 00 00 00 01 84 0a         .r............

```
01: Slave ID
03: Function code read
00: start address (high byte)
00: start address (low byte)
00: number of registers to ready (high byte)
01: number of registers to read (low byte)
84: CRC checksum
0a: CRC checksum

**Reply by janiversen** [reply]
And ?

**Reply by JimmyTang999** [reply]
and this is whats happening when i run pymodbus

```
0020   01 c3 d9 2e 00 1a 66 a1 56 6d 00 00 34 49 50 10   ......f.Vm..4IP.
0030   fa f0 85 0f 00 00                                 ......

0020   01 c3 d9 2e 00 1a 66 a1 56 6d 00 00 34 49 50 18   ......f.Vm..4IP.
0030   fa f0 85 1b 00 00 00 01 00 00 00 06 01 03 0f a0   ................
0040   00 0a    
```

pretty different structure :(

**Reply by janiversen** [reply]
Of course, you need to use rtu over tcp, the structure you see is for socket.

**Reply by JimmyTang999** [reply]
changed to

`client = ModbusTcpClient('192.168.1.195', port=26,framer=FramerType.RTU)
`
but now getting a slightly different error:
Error reading registers: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received)

**Reply by JimmyTang999** [reply]
so that got me closer
```
0020   01 c3 eb a7 00 1a 7d ce f1 bd 00 00 40 2b 50 18   ......}.....@+P.
0030   fa f0 85 17 00 00 01 03 0f a0 00 0a c6 fb         ..............
```

i noticed it closely matched radzio. 
`01 03 0f a0 00 0a c6 fb 
`
the address looked too large so, i modified the syntax
result = client.read_holding_registers(address=0, count=10, slave=1)

i guess the library is smart enough to know when you say holding_register, it automatically starts you at 4000
so if you want to start with the first holding register, specify 0 

heres my working code if it helps someone else in the future

```
from pymodbus.client import ModbusTcpClient
from pymodbus import (ExceptionResponse,FramerType,ModbusException,pymodbus_apply_logging_config)
client = ModbusTcpClient('192.168.1.195', port=26,framer=FramerType.RTU)
if client.connect():
    print("connected")
else:
    print("not connectd")
result = client.read_holding_registers(address=0, count=10, slave=1)
if result.isError():
    print("Error reading registers:", result)
else:
    print("Register values:", result.registers)
client.close()
```

**Reply by janiversen** [reply]
No pymodbus do not assume holding registers start at a specific address, because different devices uses different addresses.

Depending on how your vendor specifies the addresses is might be necessary to add +1.

address 4000 is correct as 0x0F 0xA0, but you might need to ask for 4001.

**Comment by janiversen** [comment]
The picture shows RTU over tcp/ip but your code uses socket, so maybe that is the problem,

Please refer to our documentation.
